packages/ibm-products/src/components/TagOverflow/TagOverflow.js

```diff
@@ -0,0 +1,266 @@
+// values are initialized early to avoid react making unnecessary re-renders.
+// Note that default values are not required for props that are 'required',
+// nor for props where the component can apply undefined values reasonably.
+// Default values should be provided when the component needs to make a choice
+// or assumption when a prop is not supplied.
+
+// Default values for props
+// const defaults = {
+//   /* TODO: add defaults for relevant props if needed */
+// };
+
+/**
+ * TODO: A description of the component.
+ */
+export let TagOverflow = React.forwardRef(
+  (
+    {
+      // The component props, in alphabetical order (for consistency).
+      className,
+      items,
+      itemTemplate,
+      maxVisible,
+      // Collect any other property values passed in.
+      ...rest
+    },
+    ref
+  ) => {
+    // type of Tags elements within the component
+    const tagType = 'blue';
+
+    const localContainerRef = useRef(null);
+    const containerRef = ref || localContainerRef;
+    const itemRefs = useRef(null);
+    const overflowRef = useRef(null);
+    // measurementOffset is the value of margin applied on each items,
+    // this value is required for calculating how many items will fit within the container
+    const measurementOffset = 4;
+    const overflowIndicatorWidth = 40;
+
+    const [containerWidth, setContainerWidth] = useState(0);
+    const [visibleItems, setVisibleItems] = useState([]);
+    const [overflowItems, setOverflowItems] = useState([]);
+
+    const handleResize = () => {
+      setContainerWidth(containerRef.current.offsetWidth);
+    };
+
+    useResizeObserver(containerRef, handleResize);
+
+    const getMap = () => {
+      if (!itemRefs.current) {
+        itemRefs.current = new Map();
+      }
+      return itemRefs.current;
+    };
+
+    const itemRefHandler = (id, node) => {
+      const map = getMap();
+      if (id && node && map.get(id) !== node.offsetWidth) {
+        map.set(id, node.offsetWidth);
+      }
+    };
+
+    const getVisibleItems = useCallback(() => {
+      if (!itemRefs.current) {
+        return items;
+      }
+
+      const map = getMap();
+      const overflowContanerWidth =
+        overflowRef.current.offsetWidth > overflowIndicatorWidth
+          ? overflowRef.current.offsetWidth
+          : overflowIndicatorWidth;
+      const maxWidth = containerWidth - overflowContanerWidth;
+
+      let childrenWidth = 0;
+      let maxReached = false;
+
+      return items.reduce((prev, cur) => {
+        if (!maxReached) {
+          const itemWidth = map.get(cur.id) + measurementOffset;
+          const fits = itemWidth + childrenWidth < maxWidth;
+
+          if (fits) {
+            childrenWidth += itemWidth;
+            prev.push(cur);
+          } else {
+            maxReached = true;
+          }
+        }
+        return prev;
+      }, []);
+    }, [itemRefs, overflowRef, containerWidth, items]);
+
+    const ItemComponent = ({ templateProps }) => {
```

Comment:
typically i would expect to see any other additional component in it's own separate file for organizational purposes. also i like `itemTemplate` but i think something like `tagComponent` is slightly more straightforward.

you could just keep this here as a simple function. something like

```
const getCustomComponent = (item) => {
  const { className, id, ...other } = item;
  return createElement(tagComponent, {
    ...other,
    className: cx(`${blockClass}__item`, className),
    ref: (node) => itemRefHandler(id, node),
  });
};
```

Reply:
Made the changes as suggested. Thank you for optimising it nicely.